### DISCUSSION TITLE
Newest Typescript breaks stuff

### DISCUSSION THREAD

**Discussion by marineloken** [sveltejs/kit#10695] [General]
Hello  

I have discovered a problem that could affect people using SvelteKit in larger projects:

In a SvelteKit project with:
- TypeScript
- "moduleResolution": "NodeNext" in tsconfig.json [because of svelte-package packages](https://kit.svelte.dev/docs/packaging#caveats)
Then using the latest versions for everything is fine, except for typescript itself; Installing a version beyond 5.1.6 results in:
```log
error TS5110: Option 'module' must be set to 'NodeNext' when option 'moduleResolution' is set to 'NodeNext'.
```

And setting "module": "NodeNext" in tsconfig.json introduces a host of other issues.

Anyone have experienced this or has thoughts, workarounds, etc.?

**Comment by jKm00** [comment]
Hi, I experienced a similiar issue as 'module' was set to 'NodeNext' by default when setting up a new Sveltekit library. When building the library I got this warning:

```
1. pkg.exports["."].svelte is ./dist/index.js and is written in CJS, but is interpreted as ESM. Consider using the .cjs extension, e.g. ./dist/index.cjs
```

Additionally, when trying to use the library in an application, it fails with: 'ReferenceError: exports is not defined'.

By removing the 'module' from the `ts.config`, I can build the library with no warnings and I can use it in another application. However, the `ts.config` complains with the same error you mentioned: 'Option 'module' must be set to 'NodeNext' when option 'moduleResolution' is set to 'NodeNext'.'

**Reply by marineloken** [reply]
Yes, the package exports/typescript relationship has been a bit turbulent lately, but SvelteKit has updated their [documentation](https://kit.svelte.dev/docs/packaging#typescript); Setting `"moduleResolution"` to `"bundler"` should fix it :)  
_(don't ask me why lmao)_  

Hopefully it won't change again in the near future...

**Comment by cloudymeatball** [comment]
I am having the opposite issue, where using `moduleResolution: "node16"` is better as I am using library that has node conditional exports with functionalities only for node environments. The consequences, to date, of using `moduleResolution: "bundler"` is little type safety when using such exports.
> TypeScript always matches the "types" and "default" conditions if present. [Typescript 5 documentation](https://www.typescriptlang.org/docs/handbook/modules/reference.html#the-moduleresolution-compiler-option)

This puts the onus on library maintainers to add an "types" export, which [extends the Node interface](https://nodejs.org/api/packages.html#conditional-exports). How do more experienced typescript users feel about just changing to "node16" or "nodenext" and diverge from Svelte (and Typescript) recommendations?
